### PR TITLE
encryption/decryption support in java-storage

### PR DESCRIPTION
**Pull request by singhravidutt**

### CODE REVIEW

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java [hunk 1]
```diff
@@ -494,6 +497,12 @@ private BlobSourceOption[] generateReadOptions(BlobId blobId) {
       if (blobId.getGeneration() != null) {
         readOptions.add(BlobSourceOption.generationMatch(blobId.getGeneration()));
       }
+      if (storageOptions.getEncryptionKey() != null) {
+        readOptions.add(
+            BlobSourceOption.decryptionKey(
+                checkNotNull(storageOptions.getEncryptionKey(), "encryption key must not be null")
```
[the remark this answers is not in the document]
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java:503]: `checkNotNull` is redundant here, will remove.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientWriteChannel.java [hunk 2]
```diff
@@ -156,17 +157,23 @@ public Boolean call() throws Exception {
...
     List<BlobWriteOption> writeOptions = new ArrayList<>();
 
     writeOptions.add(BlobWriteOption.disableGzipContent());
     writeOptions.add(BlobWriteOption.generationMatch());
     if (createOptions.getKmsKeyName() != null) {
       writeOptions.add(BlobWriteOption.kmsKeyName(createOptions.getKmsKeyName()));
     }
-    if (channelOptions.isGrpcChecksumsEnabled()) {
+    if (storageOptions.getWriteChannelOptions().isGrpcChecksumsEnabled()) {
       writeOptions.add(BlobWriteOption.crc32cMatch());
```
[the remark this answers is not in the document]
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientWriteChannel.java:169]: I don't think we are mutating the input here. We have created a list of `BlobWriteOption` and adding all the checks in it. Am I overlooking something?

[the remark this answers is not in the document]
Comment (arunkumarchacko) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientWriteChannel.java:168]: Added the comment in the wrong place. Please ignore this comment.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java [hunk 3]
```diff
@@ -494,6 +497,9 @@ private BlobSourceOption[] generateReadOptions(BlobId blobId) {
       if (blobId.getGeneration() != null) {
         readOptions.add(BlobSourceOption.generationMatch(blobId.getGeneration()));
       }
+      if (storageOptions.getEncryptionKey() != null) {
+        readOptions.add(BlobSourceOption.decryptionKey(storageOptions.getEncryptionKey().value()));
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java:501]: I meant here. Here we are updating the readOptions, which is an input paramter. Changing the input parameter is usually a code smell.

Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java]: We are not mutating inputParameters but the local `readOptions` variable. Refactored the name to remove the confusion.

**Review by arunkumarchacko — APPROVED** [review]

**Pull request merged**